fl4health/losses/contrastive_loss.py

```diff
@@ -86,3 +87,75 @@ def forward(
+        pairs.
+
+        Args:
+            device (torch.device): device to use for computation
+            temperature (float): temperature to scale the logits
+        """
+        super().__init__()
+        self.device = device
+        self.temperature = temperature
+
+    def forward(self, features: torch.Tensor, transformed_features: torch.Tensor) -> torch.Tensor:
+        """
+        Compute the contrastive loss based on the features and transformed_features. Given N features
+        and N transformed_features per batch, features[i] and transformed_features[i] are positive pairs
+        and the remaining 2N - 2 are negative pairs.
+
+        Args:
+            features (torch.Tensor): Features of input without transformation applied.
+                Shaped (batch_size, feature_dimension).
+            transformed_features (torch.Tensor): Features of input with transformation applied.
+                Shaped (batch_size, feature_dimension).
+
+        Returns:
+            torch.Tensor: Contrastive loss value
+        """
+
+        features.to(self.device)
+        transformed_features.to(self.device)
+
+        # Ensure features and transformed_features are same shape
+        assert features.shape == transformed_features.shape
+        batch_size = features.shape[0]
+
+        # Concatenate features and transformed features. Normalize each feature with euclidean norm.
+        all_features = torch.concatenate([features, transformed_features], dim=0).to(self.device)
+        all_features = F.normalize(all_features, dim=-1)
+
+        # Compute similarity of each features with other features
+        # Equivalent to Cosine Similarity since feature are normalized
+        similarity_matrix = torch.matmul(all_features, all_features.T)
+
+        # Extract positive pairs from similarity matrix
+        # Positive pairs are elements (i, j) offset from matrix by batch size
+        # As a result of stacking feature and transformed_features
+        similarity_ij = torch.diag(similarity_matrix, diagonal=batch_size)
+        similarity_ji = torch.diag(similarity_matrix, diagonal=-batch_size)
+        positives = torch.concatenate([similarity_ij, similarity_ji], dim=0)
+
+        # Numerator is the sum of the exponent of positive similarities
+        numerator = torch.exp(positives / self.temperature)
+
+        # Denominator is all pair combinations except for diagonal which corresponds to a features similarity to itself
+        mask = (torch.ones(2 * batch_size, 2 * batch_size) - torch.eye(2 * batch_size)).to(self.device)
+        similarity_matrix_without_diagonal = torch.mul(similarity_matrix, mask)
+        denominator = torch.exp(similarity_matrix_without_diagonal / self.temperature)
+
+        # Final loss negative log likelihood
+        losses = -torch.log(numerator / denominator.sum(dim=1))
+        loss = torch.sum(losses) / (2 * batch_size)
```

Comment:
Maybe a comment here that we're dividing by 2*batch_size since pairs are double counted due to the similarity matrix symmetry?